comp/trace/config/config.go

```diff
@@ -67,9 +75,44 @@ func NewConfig(deps Dependencies) (Component, error) {
 	}
 	c.SetMaxMemCPU(env.IsContainerized())
 
+	c.coreConfig.OnUpdate(func(setting string, oldValue, newValue any) {
+		log.Debugf("OnUpdate: %s", setting)
+		if setting != apiKeyConfigKey {
+			return
+		}
+
+		if c.coreConfig.IsSet(apmConfigAPIKeyConfigKey) {
+			// apm_config.api_key is deprecated. Since it overrides core api_key values during config setup,
+			// if used, core API Key refresh is skipped. TODO: check usage of apm_config.api_key and remove it.
+			log.Warn("cannot refresh api_key on trace-agent while `apm_config.api_key` is set. `apm_config.api_key` is deprecated, use core `api_key` instead")
+			return
+		}
+		oldAPIKey, ok1 := oldValue.(string)
+		newAPIKey, ok2 := newValue.(string)
+		if ok1 && ok2 {
+			log.Debugf("Updating API key in trace-agent config, replacing `%s` with `%s`", scrubber.HideKeyExceptLastFiveChars(oldAPIKey), scrubber.HideKeyExceptLastFiveChars(newAPIKey))
+			// Update API Key on config, and propagate the signal to registered listeners
+			newAPIKey = pkgconfigutils.SanitizeAPIKey(newAPIKey)
+			c.updateAPIKey(oldAPIKey, newAPIKey)
+		}
+	})
+
 	return &c, nil
 }
 
+func (c *cfg) updateAPIKey(oldKey, newKey string) {
+	// Update API Key on config, and propagate the signal to registered listeners
+	c.UpdateAPIKey(newKey)
+	if c.updateAPIKeyFn != nil {
+		c.updateAPIKeyFn(oldKey, newKey)
+	}
+}
+
+// OnUpdateAPIKey registers a callback for API Key changes
+func (c *cfg) OnUpdateAPIKey(callback func(oldKey, newKey string)) {
+	c.updateAPIKeyFn = callback
```

Comment:
Suggestion: log an error if `updateAPIKeyFn` is not nil on function entry, since it's not obvious from the caller's perspective that only 1 callback can be registered. Also, add a mention of this to the function's doc: `OnUpdateAPIKey registers a callback for API Key changes, only 1 callback can be used at a time`